jwst/outlier_detection/outlier_detection_ifu.py

```diff
@@ -53,203 +49,258 @@ def __init__(self, input_models, reffiles=None, **pars):
-
-        # Compute median of stack os images using BADMASKS to remove low weight
-        # values
-        median_image = median(resampled_sci, nlow=nlow, nhigh=nhigh,
-                              badmasks=badmasks)
-        # for i in range(len(resampled_sci)):
-        #    data = resampled_sci[i]
-        #    mask = badmasks[i]
-        return median_image
-
-    def blot_median(self, median_image):
-        """IFU-specific version of blot_median."""
-        cubeblot = blot_cube_build.CubeBlot(median_image, self.input_models)
-        cubeblot.blot_info()
-        blot_models, input_list_number = cubeblot.blot_images()
-        for j in range(len(blot_models)):
-            k = input_list_number[j]
-            self.blot_models[k].data += blot_models[j].data
-            self.blot_models[k].meta = blot_models[j].meta
-
-
-class ErrorWrongInstrument(Exception):
-    """ Raises an exception if the instrument is not MIRI or NIRSPEC
-    """
-    pass
+            detector_name = uq_det[idet]
+            opt_info = (kern_size[0], kern_size[1], threshold_percent,
+                        diffarr, minarr, normarr, minarr_norm)
+            opt_model = self.create_optional_results_model(opt_info)
+            opt_model.meta.filename = self.make_output_path(
+                basepath=self.input_models.meta.asn_table.products[0].name,
+                suffix=detector_name + '_outlier_output')
+            log.info("Writing out intermediate outlier file {}".format(opt_model.meta.filename))
+            opt_model.save(opt_model.meta.filename)
+
+        del diffarr
+
+        # store some information if the second flagging step is to be done. 
+        if ifu_second_check:
+            # store where the minarr is nan (neighbor pixels have nan so differences produces a nan)
+            nanminarr = np.isnan(minarr)
+            nanindx = np.where(nanminarr)
+
+        # Update DQ flag
+        for i in range(len(self.input_models)):
+            model = self.input_models[i]
+            sci = model.data
+            dq = model.dq
+
+            detector = model.meta.instrument.detector.lower()
+            # only use data from the same detector
+            if detector == uq_det[idet]:
+
+                # There could be a large number of pixels with a sci value of NaN
+                # but the dq flag of DO_NOT_USE has not been set.
+                # This can occur in Non-science regions of the detector.
+                check = np.where(
+                    np.logical_and(~np.bitwise_and(dq, dqflags.pixel['DO_NOT_USE']).astype(bool),
+                                   np.isnan(sci)))
+                log.debug("Number of pixels DQ was not set to DO_NOT_USE and Sci array was Nan{} ".
+                          format(len(check[0])))
+                # set all pixels with dq = DO_NOT_USE to have sci values of Nan
+                bad = np.where(np.bitwise_and(dq, dqflags.pixel['DO_NOT_USE']).astype(bool))
+                sci[bad] = np.nan
```

Comment:
Following Howards's question about changing the data - was this intentional?

Reply:
yes I intentionally made the science data Nan for outliers. I might have misunderstood a conversion on a tag up about this. I think it LESS confusing if the science data is flagged as a NAN if the DQ is DO_NOT _USE. That is basically what is done in cal detector 1. I also think it is less confusing to the users if BAD data has a science value of NAN. For NIRSpec there where 1000s of input data that had DQ set to DO_NOT_USE but the input data was not Nan and there were science data that was NAN but the DQ flags did not have DO_NOT_USE. Super confusing.  I first had to clean up all the input data so I was not using bad data  or a bunch of nans in the comparison.  I worked with an NIRSpec user heavily with this data (it started as a help desk ticket). I strongly recommend that if the DQ flag is set to DO_NOT_USE the Science pixel also is flagged as Nan. It is cleaner. But I will take that out if you disagree.